File: endpoint-api-version-drift/python/stripe_endpoint_version_drift.py
"""Report Stripe webhook endpoints rendering events at different api_versions.

Read only. One paginated GET and no writes: give this a RESTRICTED key with read
access to Webhook Endpoints. The repair is printed, never performed, because
this script holds a credential to a live payments account.
"""
import argparse
import logging
import os
import sys

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("stripe_endpoint_version_drift")

API = "https://api.stripe.com/v1"

# Both spellings of "this endpoint has no pin" collapse here, before anything
# deduplicates. An account with two unpinned endpoints has one shape, not two.
ACCOUNT_DEFAULT = "account default"


def normalise(api_version):
    """Map an endpoint's raw api_version onto the shape it actually renders.

    Pure. Stripe returns None on some unpinned endpoints and "" on others; both
    mean "follow the account default", so both become one sentinel.
    """
    if api_version is None or api_version == "":
        return ACCOUNT_DEFAULT
    return str(api_version)


def base_url(url):
    """The URL without its query string or fragment. Pure.

    The documented dual-endpoint upgrade puts two endpoints on one URL,
    distinguished only by a query parameter, so the query string is exactly what
    has to come off before two endpoints can be recognised as the same one.
    """
    return str(url or "").split("?", 1)[0].split("#", 1)[0]


def verdict(endpoints):
    """Classify a whole account's endpoints. Pure, so both traps are testable.

    `endpoints` is a list of dicts with `url`, `api_version` and `status`.
    Returns (state, detail).
    """
    live = [e for e in endpoints if e.get("status") == "enabled"]
    if not live:
        return ("none",
                "no enabled endpoints in this mode: nothing is being delivered, "
                "so nothing can disagree about a shape")

    versions = sorted({normalise(e.get("api_version")) for e in live})
    if len(versions) == 1:
        return ("consistent",
                "all %d enabled endpoint(s) render at %s" % (len(live), versions[0]))

    by_url = {}
    for e in live:
        by_url.setdefault(base_url(e.get("url")), set()).add(
            normalise(e.get("api_version")))
    shared = sorted(u for u, v in by_url.items() if len(v) > 1)
    if shared:
        return ("migration",
                "%d versions in use (%s), and %s is served at more than one of "
                "them. That is the dual-endpoint upgrade shape, still running: "
                "the handler is being sent every event twice, in two shapes."
                % (len(versions), ", ".join(versions), shared[0]))
    return ("drift",
            "%d versions in use (%s) across %d endpoint(s) on different URLs. "
            "The same event reaches your services in different shapes and only "
            "the ones reading a moved field will fail."
            % (len(versions), ", ".join(versions), len(live)))


def get(session, path, **params):
    r = session.get(API + path, params=params, timeout=30)
    if r.status_code == 401:
        raise SystemExit("401 from Stripe: the key is wrong, or is for the other mode")
    r.raise_for_status()
    return r.json()


def endpoints(session):
    """Every webhook endpoint in this key's mode, paginated."""
    out = []
    params = {"limit": 100}
    while True:
        page = get(session, "/webhook_endpoints", **params)
        data = page.get("data", [])
        out.extend(data)
        if not data or not page.get("has_more"):
            break
        params["starting_after"] = data[-1]["id"]
    return out


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--show-disabled", action="store_true",
                    help="also list disabled endpoints, which are never counted")
    args = ap.parse_args()

    key = os.environ.get("STRIPE_API_KEY")
    if not key:
        log.error("set STRIPE_API_KEY (use a restricted, read-only key)")
        return 2

    s = requests.Session()
    s.headers.update({"Authorization": "Bearer " + key})

    eps = endpoints(s)
    if not eps:
        log.info("no webhook endpoints configured for this key's mode")
        return 0

    state, detail = verdict(eps)
    for ep in eps:
        if ep.get("status") != "enabled" and not args.show_disabled:
            continue
        log.info("  %-9s %-24s %s", ep.get("status", "?"),
                 normalise(ep.get("api_version")), ep.get("url", "?"))

    if state in ("consistent", "none"):
        log.info("%s  %s", state, detail)
        return 0

    log.warning("%s  %s", state, detail)
    log.warning("  api_version cannot be edited, so the repair is a cutover, not "
                "an update: pick the version every consumer should be on")
    log.warning("  disable the losing endpoint: POST %s/webhook_endpoints/{id} "
                "-d disabled=true", API)
    log.warning("  once nothing depends on it, remove it: DELETE %s/"
                "webhook_endpoints/{id}", API)
    log.warning("  then pin the survivor deliberately rather than leaving it on "
                "the account default, which moves at the next upgrade")
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: endpoint-api-version-drift/node/stripe-endpoint-version-drift.mjs
/**
 * Report Stripe webhook endpoints rendering events at different api_versions.
 *
 * Read only. One paginated GET and no writes: give this a RESTRICTED key with
 * read access to Webhook Endpoints. The repair is printed, never performed.
 */
const API = 'https://api.stripe.com/v1';

// Both spellings of "this endpoint has no pin" collapse here, before anything
// deduplicates. An account with two unpinned endpoints has one shape, not two.
export const ACCOUNT_DEFAULT = 'account default';

/**
 * Map an endpoint's raw api_version onto the shape it actually renders. Pure.
 * Stripe returns null on some unpinned endpoints and '' on others.
 */
export function normalise(apiVersion) {
  if (apiVersion === null || apiVersion === undefined || apiVersion === '') {
    return ACCOUNT_DEFAULT;
  }
  return String(apiVersion);
}

/** The URL without its query string or fragment. Pure. */
export function baseUrl(url) {
  return String(url ?? '').split('?')[0].split('#')[0];
}

/**
 * Classify a whole account's endpoints. Pure, so both traps are testable.
 * `endpoints` is an array of objects with url, api_version and status.
 */
export function verdict(endpoints) {
  const live = endpoints.filter((e) => e.status === 'enabled');
  if (live.length === 0) {
    return ['none',
      'no enabled endpoints in this mode: nothing is being delivered, so ' +
      'nothing can disagree about a shape'];
  }

  const versions = [...new Set(live.map((e) => normalise(e.api_version)))].sort();
  if (versions.length === 1) {
    return ['consistent',
      `all ${live.length} enabled endpoint(s) render at ${versions[0]}`];
  }

  const byUrl = new Map();
  for (const e of live) {
    const u = baseUrl(e.url);
    if (!byUrl.has(u)) byUrl.set(u, new Set());
    byUrl.get(u).add(normalise(e.api_version));
  }
  const shared = [...byUrl.entries()].filter(([, v]) => v.size > 1)
    .map(([u]) => u).sort();
  if (shared.length > 0) {
    return ['migration',
      `${versions.length} versions in use (${versions.join(', ')}), and ` +
      `${shared[0]} is served at more than one of them. That is the ` +
      'dual-endpoint upgrade shape, still running: the handler is being sent ' +
      'every event twice, in two shapes.'];
  }
  return ['drift',
    `${versions.length} versions in use (${versions.join(', ')}) across ` +
    `${live.length} endpoint(s) on different URLs. The same event reaches your ` +
    'services in different shapes and only the ones reading a moved field will fail.'];
}

async function get(key, path, params = {}) {
  const url = new URL(API + path);
  for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);
  const res = await fetch(url, { headers: { Authorization: `Bearer ${key}` } });
  if (res.status === 401) {
    throw new Error('401 from Stripe: the key is wrong, or is for the other mode');
  }
  if (!res.ok) throw new Error(`${res.status} from ${url.pathname}`);
  return res.json();
}

export async function endpoints(key) {
  const out = [];
  const params = { limit: 100 };
  for (;;) {
    const page = await get(key, '/webhook_endpoints', params);
    const data = page.data ?? [];
    out.push(...data);
    if (data.length === 0 || !page.has_more) break;
    params.starting_after = data[data.length - 1].id;
  }
  return out;
}

async function main() {
  const key = (process.env.STRIPE_API_KEY || "dummy-stripe-api-key");
  if (!key) {
    console.error('set STRIPE_API_KEY (use a restricted, read-only key)');
    process.exitCode = 2;
    return;
  }

  const eps = await endpoints(key);
  if (eps.length === 0) {
    console.log("no webhook endpoints configured for this key's mode");
    return;
  }

  const [state, detail] = verdict(eps);
  for (const ep of eps) {
    if (ep.status !== 'enabled') continue;
    console.log(`  ${String(ep.status).padEnd(9)} ` +
                `${normalise(ep.api_version).padEnd(24)} ${ep.url ?? '?'}`);
  }

  if (state === 'consistent' || state === 'none') {
    console.log(`${state}  ${detail}`);
    return;
  }

  console.warn(`${state}  ${detail}`);
  console.warn('  api_version cannot be edited, so the repair is a cutover, not ' +
               'an update: pick the version every consumer should be on');
  console.warn(`  disable the losing endpoint: POST ${API}/webhook_endpoints/{id} ` +
               '-d disabled=true');
  console.warn(`  once nothing depends on it, remove it: DELETE ${API}/` +
               'webhook_endpoints/{id}');
  console.warn('  then pin the survivor deliberately rather than leaving it on ' +
               'the account default, which moves at the next upgrade');
  process.exitCode = 1;
}

// Only run when invoked directly. The test file imports this module, and without
// the guard main() would run there too, fail on the missing key, and set a
// non-zero exit code that fails the whole test file even as every test passes.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}
